flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/EnumerationHistory.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.enumerator;
+
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.flink.calcite.shaded.com.google.common.collect.EvictingQueue;
+
+/**
+ * This enumeration history is used for split discovery throttling. It wraps Guava {@link
+ * EvictingQueue} to provide thread safety.
+ */
+@ThreadSafe
+class EnumerationHistory {
+
+  // EvictingQueue is not thread safe.
+  private final EvictingQueue<Integer> enumerationSplitCountHistory;
+
+  EnumerationHistory(int maxHistorySize) {
+    this.enumerationSplitCountHistory = EvictingQueue.create(maxHistorySize);
+  }
+
+  /** Add the split count from the last enumeration result. */
+  synchronized void add(int splitCount) {
+    enumerationSplitCountHistory.add(splitCount);
+  }
+
+  /** @return true if split discovery should pause because assigner has too many splits already. */
+  synchronized boolean shouldPauseSplitDiscovery(int pendingSplitCountFromAssigner) {
+    if (enumerationSplitCountHistory.remainingCapacity() > 0) {
+      // only check throttling when full history is obtained.
```

Comment:
I am taking this simple approach now. I am open to checkpoint the enumeration history as part of the enumerator state if people think it is better.

Reply:
I prefer the current no-checkpoint approach. To some extent, an empty history seems to be beneficial for quickly processing the backlog of data when the application restarts?

Reply:
well, there is still subjective to the `maxPlanningSnapshotCount` limit. its default is `Integer.MAX_VALUE`. For streaming read, I actually think it should be set to some reasonable value like 6 or 10 to avoid potentially discovering too many splits into enumerator memory.

Reply:
Do we checkpoint the already planned splits?

If so then after a recovery:
- pendingSplitCountFromAssigner - returns the valid value
- totalSplitCountFromRecentDiscovery - is 0 after a recovery

If my reasoning above is right then `shouldPauseSplitDiscovery` will return `true` until there is no split left pending which might cause a small hickup after a checkpoint recovery.

Reply:
@pvary you are correct here. This is the downside of not checkpointing enumeration history. After checkpoint recovery, enumerator can discover more splits because totalSplitCountFromRecentDiscovery is 0 after recovery. There are two scenarios.
1. job was stuck in a failure/restart loop. this is not a concern as there is no successful checkpoint. newly discovered splits aren't saved.
2. here is a pathological scenario. Job was in a restart loop with successful checkpoints (like 1 or 2) btw restart. This would essentially bypass the throttling check. That is the small concern I has regarding not checkpointing. However, in practice, this scenario is very unusual. Hence, I didn't implement the checkpointing of enumeration history (for throttling purpose).

On the other hand, it's not too difficult to add the checkpointing of enumeration history for completeness or ability to handle the above edge case. Hence I am open to add it if desired.

Reply:
We might have a big file where the read takes a while (or short checkpointing period), and we face an error at the end of the file.

If it is not too much effort/overhead, I would add the checkpointing here.

Reply:
> We might have a big file where the read takes a while (or short checkpointing period), and we face an error at the end of the file

This is orthogonal to the split discovery throttling feature in this PR. It is about processing semantics / duplicates upon failure recovery. Flink can achieve exactly-once processing semantics regarding state consistency. Whether it is e2e exactly-once semantics depends on sink (transactional or not). For Iceberg sink (with transactional commit), recovery won't cause duplicates in the sink Iceberg table.

> here is a pathological scenario. Job was in a restart loop with successful checkpoints (like 1 or 2) btw restart. This would essentially bypass the throttling check. Potentially the tracked splits can keep growing. That is the small concern I has regarding not checkpointing. The concern is small, because this scenario is very unusual. Hence, I didn't implement the checkpointing of enumeration history (for throttling purpose).

Whether we checkpoint the enumeration history or not really only affects the throttling feature in this PR. I can only think of the above pathological scenario where not checkpointing can be a potential problem. But again, it is a unusual scenario but it could happen in theory.

We can add the checkpointing here without breaking the compatibility as the `IcebergEnumeratorStateSerializer` is versioned. We can just bump up the version and handle both the old and new versions.

Reply:
My example for the big file was my stab at finding another edge case where we end up infinite amount of enumerated splits. Since we do not have a state, any issue which causes restarts, but creates checkpoints could cause issues. Slow and failing reads could be one such example.

Reply:
@pvary Just to make sure that I understand you correctly. 

You agree that the big file and restart is not a problem for causing infinite amount of enumerated splits. 

You see that constant restarts with completed checkpoints btw can lead to ever-growing enumerated splits. For that reason, you are in favor of adding the checkpointing of the enumeration history for more proper throttling behavior, right?

Reply:
What I missed before, but realized now is that the reader stores the row position of the file reader in the checkpoint state. So the reader will not start from the beginning, and this way we will not advance the checkpoint infinitely...

Still feel that it will cause issues/misunderstanding later if only some part of the IcebergSource state is checkpointed. The source should be either fully checkpointed or not at all.
I would not have exceptions for this rule without a very strong reason.

Reply:
fair enough. added the checkpoint of the enumeration split count history (for throttling purpose)

Reply:
Thx